File: repeatingSubstringPattern/repeatingPattern.js
// Given a string, check if it can be constructed by taking a substring of it and appending multiple copies of the substring together.
// Javascript implementation

function repPattern(s){
    let i, j = 0;
    let intHalf = (s.length - (s.length % 2)) / 2;
    let subPat = s.slice(0,intHalf);

    if (s.length === 1) {return true};

    for (i = intHalf; i >= 0; i--) {
        if (s.length % subPat.length == 0) {
            while (j < s.length && s[j] == subPat[j % subPat.length]) {
                j++;
            }
            if (j == s.length) {return true}
        }
        j = 0;
        subPat = s.slice(0,i);
    }

    return false;
}

console.log(repPattern("abcabcabc"));

File: repeatingSubstringPattern/repeatingPattern.py
# Given a string, check if it can be constructed by taking a substring of it and appending multiple copies of the substring together.
# Python implementation

def repPattern(s):
    half = len(s) // 2
    subPat = s[0:half]
    j = 0

    if len(s) == 1: return True

    for i in range(half, 1, -1):
        if len(s) % len(subPat) == 0:
            while ( j < len(s) and s[j] == subPat[j % len(subPat)]):
                j += 1
            if (j == len(s)): return True
        j = 0
        subPat = s[0:i]

    return False

print(repPattern("abcabcabc"))